commands/qemu-guest-tools/actions_test.go

```diff
@@ -0,0 +1,261 @@
+	nilOrPanic(err, "Failed to create test-folder/")
+
+	testFile2 := filepath.Join(testFolder, "hello2.txt")
+	err = ioutil.WriteFile(testFile2, []byte("hello-world-2"), 0777)
+	nilOrPanic(err, "Failed to create testFile2: ", testFile2)
+
+	debug(" - meta.ListFolder")
+	files, err := meta.ListFolder(f.Path())
+	nilOrPanic(err, "ListFolder failed, err: ", err)
+
+	assert(len(files) == 2, "Expected 2 files")
+	assert(files[0] == testFile || files[1] == testFile, "Expected testFile")
+	assert(files[0] == testFile2 || files[1] == testFile2, "Expected testFile2")
+
+	////////////////////
+	debug("### Test meta.ListFolder (missing folder)")
+	files, err = meta.ListFolder(filepath.Join(f.Path(), "no-such-folder"))
+	assert(files == nil, "Expected files == nil, we hopefully have an error")
+	assert(err == engines.ErrResourceNotFound, "Expected ErrResourceNotFound")
+
+	////////////////////
+	debug("### Test meta.ListFolder (empty folder)")
+	emptyFolder := filepath.Join(f.Path(), "empty-folder")
+	err = os.Mkdir(emptyFolder, 0777)
+	nilOrPanic(err, "Failed to create empty-folder/")
+
+	files, err = meta.ListFolder(emptyFolder)
+	assert(len(files) == 0, "Expected zero files")
+	assert(err == nil, "Didn't expect any error")
+
+	////////////////////
+	testShellHello(meta)
+	testShellCat(meta)
+	testShellCatStdErr(meta)
+}
+
+func testShellHello(meta *metaservice.MetaService) {
+	debug("### Test meta.Shell (using 'echo hello')")
+	shell, err := meta.ExecShell()
+	nilOrPanic(err, "Failed to call meta.ExecShell()")
+
+	readHello := sync.WaitGroup{}
+	readHello.Add(1)
+	// Discard stderr
+	go io.Copy(ioutil.Discard, shell.StderrPipe())
+	go func() {
+		shell.StdinPipe().Write([]byte("echo HELLO\n"))
+		readHello.Wait()
+		shell.StdinPipe().Close()
+	}()
+	go func() {
+		data := bytes.Buffer{}
+		for {
+			b := []byte{0}
+			n, werr := shell.StdoutPipe().Read(b)
+			data.Write(b[:n])
+			if strings.Contains(data.String(), "HELLO") {
+				readHello.Done()
+				break
+			}
+			if werr != nil {
+				assert(werr == io.EOF, "Expected EOF!")
+				break
+			}
+		}
+		// Discard the rest
+		go io.Copy(ioutil.Discard, shell.StdoutPipe())
+	}()
+
+	success, err := shell.Wait()
+	nilOrPanic(err, "Got an error from shell.Wait, error: ", err)
+	assert(success, "Expected success from shell, we closed with end of stdin")
+}
+
+func testShellCat(meta *metaservice.MetaService) {
+	debug("### Test meta.Shell (using 'exec cat -')")
+	shell, err := meta.ExecShell()
+	nilOrPanic(err, "Failed to call meta.ExecShell()")
+
+	input := make([]byte, 42*1024*1024+7)
+	rand.Read(input)
+
+	// Discard stderr
+	go io.Copy(ioutil.Discard, shell.StderrPipe())
+	go func() {
+		if goruntime.GOOS == "windows" {
+			shell.StdinPipe().Write([]byte("type con\n"))
+		} else {
+			shell.StdinPipe().Write([]byte("exec cat -\n"))
+		}
+		// Give cat - some time to start, bash/busybox/dash won't work otherwise
+		// Or they will work, but only intermittently!!!
+		time.Sleep(250 * time.Millisecond)
```

Comment:
I boot a virtual machine like 10 times during testing... 250ms is nothing :) hehe